udapi/block/tokenize/onwhitespace.py

```diff
@@ -40,8 +76,28 @@ def process_tree(self, root):
                 # $sentence = $rest if (defined $rest);
                 raise ValueError('tokenization does not match: "%s" vs "%s"' % (token, sentence))
 
+            # Delete the token from the begining of the sentence.
+            sentence = sentence[len(token):]
+
+            # Set the SpaceAfter and SpacesAfter properly
+            m = re.match(r'\s+', sentence)
+            if m is not None:
+                spaces_after = m.group(0)
+                sentence = sentence[len(spaces_after):]
+
+            # normalize whitespace
+            if self.normalize_spaces:
+                spaces_before = ""
+                # spaces_after = "" <=> SpaceAfter=No is never set for the last token <=> len(sentence) = 0
+                spaces_after = "" if not len(spaces_after) and len(sentence) else " "
+
             # create a new node
             node = root.create_child(form=token)
             node.ord = i
-            if not space_after:
-                node.misc = 'SpaceAfter=No'
+
+            if i == 1 and len(spaces_before) > 0:
```

Comment:
`if i == 1 and spaces_before:` is more Pythonic and faster. Note that `bool(' ') is True`.